// @redhat-cloud-services/frontend-components/ConditionalFilter/TextFilter.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var prop_types_1 = __importDefault(require("prop-types"));
var react_core_1 = require("@patternfly/react-core");
var react_icons_1 = require("@patternfly/react-icons");
require("./conditional-filter.css");
var Text = (function (_super) {
    __extends(Text, _super);
    function Text() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            stateValue: '',
        };
        _this.onChangeValue = function (e, value) {
            _this.setState({
                stateValue: value,
            });
        };
        return _this;
    }
    Text.prototype.render = function () {
        var _a = this.props, value = _a.value, onChange = _a.onChange, onSubmit = _a.onSubmit, id = _a.id, icon = _a.icon, className = _a.className, isDisabled = _a.isDisabled, props = __rest(_a, ["value", "onChange", "onSubmit", "id", "icon", "className", "isDisabled"]);
        var Icon = icon || react_icons_1.SearchIcon;
        var stateValue = this.state.stateValue;
        var changeCallback = onChange ? onChange : this.onChangeValue;
        return (react_1.default.createElement(react_1.Fragment, null,
            react_1.default.createElement(react_core_1.TextInput, __assign({}, props, { className: "ins-c-conditional-filter ".concat(className || ''), id: id, isDisabled: isDisabled, value: onChange ? value : stateValue, onChange: function (_inputValue, e) { return changeCallback(e, e.target.value); }, "widget-type": "InsightsInput", onKeyDown: function (e) { return e.key === 'Enter' && onSubmit(e, value || stateValue); }, "data-ouia-component-type": "PF4/TextInput", ouiaId: "ConditionalFilter" })),
            react_1.default.createElement(Icon, { size: "sm", className: "ins-c-search-icon" })));
    };
    return Text;
}(react_1.Component));
Text.propTypes = {
    value: prop_types_1.default.string,
    placeholder: prop_types_1.default.string,
    onChange: prop_types_1.default.func,
    onSubmit: prop_types_1.default.func,
    isDisabled: prop_types_1.default.bool,
    'aria-label': prop_types_1.default.string,
    id: prop_types_1.default.string,
    icon: prop_types_1.default.elementType,
    className: prop_types_1.default.string,
};
Text.defaultProps = {
    value: '',
    onSubmit: function () { return undefined; },
    isDisabled: false,
    'aria-label': 'text input',
};
exports.default = Text;
