// @redhat-cloud-services/frontend-components-utilities/inventoryDependencies/inventoryDependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reactCore = exports.reactIcons = void 0;
var react_core_1 = require("@patternfly/react-core");
var react_icons_1 = require("@patternfly/react-icons");
exports.reactIcons = {
    CubesIcon: react_icons_1.CubesIcon,
    ExclamationTriangleIcon: react_icons_1.ExclamationTriangleIcon,
};
exports.reactCore = {
    Badge: react_core_1.Badge,
    Bullseye: react_core_1.Bullseye,
    Button: react_core_1.Button,
    Checkbox: react_core_1.Checkbox,
    ClipboardCopy: react_core_1.ClipboardCopy,
    Dropdown: react_core_1.Dropdown,
    DropdownItem: react_core_1.DropdownItem,
    DropdownPosition: react_core_1.DropdownPosition,
    DropdownToggle: react_core_1.DropdownToggle,
    EmptyState: react_core_1.EmptyState,
    EmptyStateBody: react_core_1.EmptyStateBody,
    EmptyStateIcon: react_core_1.EmptyStateIcon,
    EmptyStateVariant: react_core_1.EmptyStateVariant,
    Flex: react_core_1.Flex,
    FlexItem: react_core_1.FlexItem,
    Grid: react_core_1.Grid,
    GridItem: react_core_1.GridItem,
    KebabToggle: react_core_1.KebabToggle,
    Level: react_core_1.Level,
    LevelItem: react_core_1.LevelItem,
    Modal: react_core_1.Modal,
    Pagination: react_core_1.Pagination,
    PaginationVariant: react_core_1.PaginationVariant,
    Split: react_core_1.Split,
    SplitItem: react_core_1.SplitItem,
    Stack: react_core_1.Stack,
    StackItem: react_core_1.StackItem,
    Spinner: react_core_1.Spinner,
    Tab: react_core_1.Tab,
    Tabs: react_core_1.Tabs,
    Title: react_core_1.Title,
    Tooltip: react_core_1.Tooltip,
    Toolbar: react_core_1.Toolbar,
    ToolbarItem: react_core_1.ToolbarItem,
    ToolbarContent: react_core_1.ToolbarContent,
    ToolbarGroup: react_core_1.ToolbarGroup,
    TextInput: react_core_1.TextInput,
    ChipGroup: react_core_1.ChipGroup,
    Chip: react_core_1.Chip,
    DropdownToggleCheckbox: react_core_1.DropdownToggleCheckbox,
    Select: react_core_1.Select,
    SelectOption: react_core_1.SelectOption,
    SelectVariant: react_core_1.SelectVariant,
    SelectGroup: react_core_1.SelectGroup,
    Radio: react_core_1.Radio,
    Drawer: react_core_1.Drawer,
    DrawerPanelContent: react_core_1.DrawerPanelContent,
    DrawerContent: react_core_1.DrawerContent,
    DrawerContentBody: react_core_1.DrawerContentBody,
    DrawerPanelBody: react_core_1.DrawerPanelBody,
    DrawerActions: react_core_1.DrawerActions,
    DrawerHead: react_core_1.DrawerHead,
    DrawerCloseButton: react_core_1.DrawerCloseButton,
    Switch: react_core_1.Switch,
    Form: react_core_1.Form,
    Label: react_core_1.Label,
    FormGroup: react_core_1.FormGroup,
    FormSelect: react_core_1.FormSelect,
    FormSelectOption: react_core_1.FormSelectOption,
};
