// @redhat-cloud-services/frontend-components-utilities/esm/inventoryDependencies/inventoryDependencies.js
import { Badge, Bullseye, Button, Checkbox, ClipboardCopy, Dropdown, DropdownItem, DropdownPosition, DropdownToggle, DropdownToggleCheckbox, TextInput, Grid, GridItem, Tab, Tabs, Switch, Form, FormGroup, FormSelect, FormSelectOption, Label, EmptyState, EmptyStateBody, EmptyStateIcon, EmptyStateVariant, Select, SelectOption, SelectVariant, SelectGroup, Radio, ChipGroup, Chip, Title, Split, SplitItem, Spinner, Flex, FlexItem, Pagination, PaginationVariant, Tooltip, Modal, Stack, StackItem, Level, LevelItem, KebabToggle, Toolbar, ToolbarItem, ToolbarContent, ToolbarGroup, Drawer, DrawerPanelContent, DrawerContent, DrawerContentBody, DrawerPanelBody, DrawerActions, DrawerHead, DrawerCloseButton, } from '@patternfly/react-core';
import { CubesIcon, ExclamationTriangleIcon } from '@patternfly/react-icons';
export var reactIcons = {
    CubesIcon: CubesIcon,
    ExclamationTriangleIcon: ExclamationTriangleIcon,
};
export var reactCore = {
    Badge: Badge,
    Bullseye: Bullseye,
    Button: Button,
    Checkbox: Checkbox,
    ClipboardCopy: ClipboardCopy,
    Dropdown: Dropdown,
    DropdownItem: DropdownItem,
    DropdownPosition: DropdownPosition,
    DropdownToggle: DropdownToggle,
    EmptyState: EmptyState,
    EmptyStateBody: EmptyStateBody,
    EmptyStateIcon: EmptyStateIcon,
    EmptyStateVariant: EmptyStateVariant,
    Flex: Flex,
    FlexItem: FlexItem,
    Grid: Grid,
    GridItem: GridItem,
    KebabToggle: KebabToggle,
    Level: Level,
    LevelItem: LevelItem,
    Modal: Modal,
    Pagination: Pagination,
    PaginationVariant: PaginationVariant,
    Split: Split,
    SplitItem: SplitItem,
    Stack: Stack,
    StackItem: StackItem,
    Spinner: Spinner,
    Tab: Tab,
    Tabs: Tabs,
    Title: Title,
    Tooltip: Tooltip,
    Toolbar: Toolbar,
    ToolbarItem: ToolbarItem,
    ToolbarContent: ToolbarContent,
    ToolbarGroup: ToolbarGroup,
    TextInput: TextInput,
    ChipGroup: ChipGroup,
    Chip: Chip,
    DropdownToggleCheckbox: DropdownToggleCheckbox,
    Select: Select,
    SelectOption: SelectOption,
    SelectVariant: SelectVariant,
    SelectGroup: SelectGroup,
    Radio: Radio,
    Drawer: Drawer,
    DrawerPanelContent: DrawerPanelContent,
    DrawerContent: DrawerContent,
    DrawerContentBody: DrawerContentBody,
    DrawerPanelBody: DrawerPanelBody,
    DrawerActions: DrawerActions,
    DrawerHead: DrawerHead,
    DrawerCloseButton: DrawerCloseButton,
    Switch: Switch,
    Form: Form,
    Label: Label,
    FormGroup: FormGroup,
    FormSelect: FormSelect,
    FormSelectOption: FormSelectOption,
};
