// @redhat-cloud-services/frontend-components/PrimaryToolbar/Actions.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.actionPropsGenerator = exports.overflowActionsMapper = void 0;
var react_1 = __importStar(require("react"));
var react_core_1 = require("@patternfly/react-core");
var prop_types_1 = __importDefault(require("prop-types"));
var DownloadButton_1 = require("../DownloadButton");
var overflowActionsMapper = function (action, key) { return (react_1.default.createElement(react_core_1.DropdownItem, __assign({}, action.props, { className: "ins-c-primary-toolbar__overflow-actions", key: action.value || action.key || "".concat(key, "-overflow"), component: (action.props && action.props.component) || react_1.default.isValidElement(action.label || action) ? 'div' : 'button', onClick: function (e) { return action.onClick && action.onClick(e, action, key); } }), action.label || action)); };
exports.overflowActionsMapper = overflowActionsMapper;
var actionPropsGenerator = function (action, key) { return (__assign(__assign({}, action.props), { component: (action.props && action.props.component) || (react_1.default.isValidElement(action.label || action) ? 'div' : 'button'), onClick: function (e) { return action.onClick && action.onClick(e, action, key); }, children: action.label || action })); };
exports.actionPropsGenerator = actionPropsGenerator;
var Actions = (function (_super) {
    __extends(Actions, _super);
    function Actions() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            isOpen: false,
        };
        _this.toggleOpen = function (isOpen) {
            _this.setState({ isOpen: isOpen });
        };
        return _this;
    }
    Actions.prototype.render = function () {
        var _this = this;
        var isOpen = this.state.isOpen;
        var _a = this.props, actions = _a.actions, overflowActions = _a.overflowActions, onSelect = _a.onSelect, dropdownProps = _a.dropdownProps, kebabToggleProps = _a.kebabToggleProps, exportConfig = _a.exportConfig;
        var firstAction = actions[0], restActions = actions.slice(1);
        return (react_1.default.createElement(react_1.Fragment, null,
            firstAction && (react_1.default.createElement(react_core_1.ToolbarItem, { className: "ins-c-primary-toolbar__first-action pf-m-spacer-sm" }, firstAction.label ? (react_1.default.createElement(react_core_1.Button, __assign({ ouiaId: "".concat(firstAction.label) }, firstAction.props, { onClick: firstAction.onClick || (firstAction.props && firstAction.props.onClick) || undefined }), firstAction.label)) : (firstAction))),
            exportConfig && (exportConfig.extraItems || exportConfig.onSelect) && (react_1.default.createElement(react_core_1.ToolbarItem, { className: "pf-m-spacer-sm" },
                react_1.default.createElement(DownloadButton_1.DownloadButton, __assign({}, exportConfig)))),
            ((actions && actions.length > 0) || overflowActions.length > 0) && (react_1.default.createElement(react_core_1.ToolbarItem, { className: "".concat(actions.length <= 1 ? 'ins-m-actions--empty' : '', " ins-c-primary-toolbar__actions pf-m-spacer-sm") },
                react_1.default.createElement(react_core_1.Dropdown, __assign({}, dropdownProps, { isOpen: isOpen, isPlain: true, onSelect: function () {
                        var props = [];
                        for (var _i = 0; _i < arguments.length; _i++) {
                            props[_i] = arguments[_i];
                        }
                        onSelect && onSelect.apply(void 0, props);
                        _this.toggleOpen(false);
                    }, ouiaId: "Actions", toggle: react_1.default.createElement(react_core_1.KebabToggle, __assign({}, kebabToggleProps, { onToggle: function (isOpen) { return _this.toggleOpen(isOpen); } })), dropdownItems: __spreadArray(__spreadArray(__spreadArray(__spreadArray([], (firstAction
                        ? [
                            react_1.default.createElement(react_core_1.DropdownItem, __assign({ key: "first-action" }, (0, exports.actionPropsGenerator)(firstAction, 'first-action'), { className: "ins-c-primary-toolbar__first-action ".concat((firstAction.props && firstAction.props.className) || '') })),
                        ]
                        : []), true), restActions.map(function (action, key) { return (react_1.default.createElement(react_core_1.DropdownItem, __assign({ key: action.key || (action && action.props && action.props.key) || key }, (0, exports.actionPropsGenerator)(action, key)))); }), true), (actions.length > 0 && overflowActions.length > 0
                        ? [react_1.default.createElement(react_core_1.DropdownSeparator, { key: "separator", className: "ins-c-primary-toolbar__overflow-actions-separator" })]
                        : []), true), overflowActions.map(exports.overflowActionsMapper), true) }))))));
    };
    return Actions;
}(react_1.Component));
var actionsType = prop_types_1.default.arrayOf(prop_types_1.default.oneOfType([
    prop_types_1.default.node,
    prop_types_1.default.shape({
        label: prop_types_1.default.node,
        value: prop_types_1.default.oneOfType([prop_types_1.default.number, prop_types_1.default.string]),
        onClick: prop_types_1.default.func,
        props: prop_types_1.default.any,
    }),
    prop_types_1.default.string,
]));
Actions.propTypes = {
    actions: actionsType,
    onSelect: prop_types_1.default.func,
    overflowActions: actionsType,
    dropdownProps: prop_types_1.default.shape((_a = {},
        _a[prop_types_1.default.string] = prop_types_1.default.any,
        _a)),
    kebabToggleProps: prop_types_1.default.shape((_b = {},
        _b[prop_types_1.default.string] = prop_types_1.default.any,
        _b)),
    exportConfig: prop_types_1.default.shape(DownloadButton_1.DownloadButton.propTypes),
};
Actions.defaultProps = {
    actions: [],
    overflowActions: [],
    dropdownProps: {},
    exportConfig: {},
    onSelect: function () { return undefined; },
};
exports.default = Actions;
