// @redhat-cloud-services/frontend-components/PrimaryToolbar/PrimaryToolbar.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
var react_core_1 = require("@patternfly/react-core");
var react_icons_1 = require("@patternfly/react-icons");
var react_1 = __importStar(require("react"));
var Actions_1 = __importDefault(require("./Actions"));
var BulkSelect_1 = require("../BulkSelect");
var ConditionalFilter_1 = require("../ConditionalFilter");
var DownloadButton_1 = require("../DownloadButton");
var FilterChips_1 = require("../FilterChips");
var prop_types_1 = __importDefault(require("prop-types"));
var SortBy_1 = __importDefault(require("./SortBy"));
var react_table_1 = require("@patternfly/react-table");
require("./primary-toolbar.css");
var PrimaryToolbar = (function (_super) {
    __extends(PrimaryToolbar, _super);
    function PrimaryToolbar() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    PrimaryToolbar.prototype.render = function () {
        var _a = this.props, id = _a.id, className = _a.className, toggleIsExpanded = _a.toggleIsExpanded, bulkSelect = _a.bulkSelect, filterConfig = _a.filterConfig, dedicatedAction = _a.dedicatedAction, actionsConfig = _a.actionsConfig, sortByConfig = _a.sortByConfig, pagination = _a.pagination, activeFiltersConfig = _a.activeFiltersConfig, children = _a.children, exportConfig = _a.exportConfig, expandAll = _a.expandAll, useMobileLayout = _a.useMobileLayout, props = __rest(_a, ["id", "className", "toggleIsExpanded", "bulkSelect", "filterConfig", "dedicatedAction", "actionsConfig", "sortByConfig", "pagination", "activeFiltersConfig", "children", "exportConfig", "expandAll", "useMobileLayout"]);
        var overflowActions = __spreadArray([], (sortByConfig
            ? [
                {
                    label: 'Sort order ASC',
                    props: { isDisabled: sortByConfig.direction === react_table_1.SortByDirection.asc },
                    onClick: function (e) { return sortByConfig.onSortChange && sortByConfig.onSortChange(e, react_table_1.SortByDirection.asc); },
                },
                {
                    label: 'Sort order DESC',
                    props: { isDisabled: sortByConfig.direction === react_table_1.SortByDirection.desc },
                    onClick: function (e) { return sortByConfig.onSortChange && sortByConfig.onSortChange(e, react_table_1.SortByDirection.desc); },
                },
            ]
            : []), true);
        return (react_1.default.createElement(react_core_1.Toolbar, __assign({}, props, { className: "".concat(className || '', " ins-c-primary-toolbar"), toggleIsExpanded: toggleIsExpanded, id: id || 'ins-primary-data-toolbar', ouiaId: "PrimaryToolbar" }),
            react_1.default.createElement(react_core_1.ToolbarContent, null,
                (expandAll || bulkSelect || filterConfig || dedicatedAction) && (react_1.default.createElement(react_core_1.ToolbarGroup, { className: "ins-c-primary-toolbar__group-filter pf-m-spacer-md pf-m-space-items-lg", variant: "filter-group" },
                    expandAll && (react_1.default.createElement(react_core_1.ToolbarItem, null, react_1.default.isValidElement(expandAll) ? (expandAll) : (react_1.default.createElement(react_core_1.Button, __assign({ variant: "plain", "aria-label": "".concat(expandAll.isAllExpanded ? 'Collapse' : 'Expand', " all"), onClick: function (e) { return expandAll.onClick(e, !expandAll.isAllExpanded); }, ouiaId: "ExpandCollapseAll", isDisabled: expandAll.isDisabled }, expandAll.buttonProps),
                        react_1.default.createElement(react_core_1.ToolbarExpandIconWrapper, null, expandAll.isAllExpanded ? react_1.default.createElement(react_icons_1.AngleDownIcon, null) : react_1.default.createElement(react_icons_1.AngleRightIcon, null)))))),
                    bulkSelect && (react_1.default.createElement(react_core_1.ToolbarItem, null, react_1.default.isValidElement(bulkSelect) ? bulkSelect : react_1.default.createElement(BulkSelect_1.BulkSelect, __assign({ ouiaId: "BulkSelect" }, bulkSelect)))),
                    filterConfig && (react_1.default.createElement(react_core_1.ToolbarItem, { className: "ins-c-primary-toolbar__filter" }, react_1.default.isValidElement(filterConfig) ? filterConfig : react_1.default.createElement(ConditionalFilter_1.ConditionalFilter, __assign({ useMobileLayout: useMobileLayout }, filterConfig)))),
                    dedicatedAction && react_1.default.createElement(react_core_1.ToolbarItem, null, dedicatedAction))),
                react_1.default.isValidElement(actionsConfig)
                    ? actionsConfig
                    : ((actionsConfig && actionsConfig.actions && actionsConfig.actions.length > 0) || sortByConfig || exportConfig) && (react_1.default.createElement(Actions_1.default, __assign({}, (actionsConfig || {}), { exportConfig: exportConfig, overflowActions: overflowActions }))),
                sortByConfig && (react_1.default.createElement(react_core_1.ToolbarItem, { className: "ins-c-primary-toolbar__sort-by" }, react_1.default.isValidElement(sortByConfig) ? sortByConfig : react_1.default.createElement(SortBy_1.default, __assign({}, sortByConfig)))),
                children,
                pagination && (react_1.default.createElement(react_core_1.ToolbarItem, { className: "ins-c-primary-toolbar__pagination" }, react_1.default.isValidElement(pagination) ? pagination : react_1.default.createElement(react_core_1.Pagination, __assign({ isCompact: true, ouiaId: "CompactPagination" }, pagination))))),
            activeFiltersConfig && react_1.default.isValidElement(activeFiltersConfig) ? (react_1.default.createElement(react_core_1.ToolbarContent, null,
                react_1.default.createElement(react_core_1.ToolbarItem, null, activeFiltersConfig))) : (activeFiltersConfig !== undefined &&
                (activeFiltersConfig.filters.length > 0 || activeFiltersConfig.showDeleteButton === true) && (react_1.default.createElement(react_core_1.ToolbarContent, null,
                react_1.default.createElement(react_core_1.ToolbarItem, null,
                    react_1.default.createElement(FilterChips_1.FilterChips, __assign({}, activeFiltersConfig))))))));
    };
    return PrimaryToolbar;
}(react_1.Component));
PrimaryToolbar.propTypes = {
    id: prop_types_1.default.oneOfType([prop_types_1.default.number, prop_types_1.default.string]),
    className: prop_types_1.default.string,
    toggleIsExpanded: prop_types_1.default.func,
    bulkSelect: prop_types_1.default.shape(BulkSelect_1.BulkSelect.propTypes),
    filterConfig: prop_types_1.default.shape(ConditionalFilter_1.ConditionalFilter.propTypes),
    dedicatedAction: prop_types_1.default.node,
    pagination: prop_types_1.default.shape(react_core_1.Pagination.propTypes),
    sortByConfig: prop_types_1.default.shape(SortBy_1.default.propTypes),
    exportConfig: prop_types_1.default.shape(DownloadButton_1.DownloadButton.propTypes),
    activeFiltersConfig: prop_types_1.default.shape(FilterChips_1.FilterChips.propTypes),
    children: prop_types_1.default.node,
    actionsConfig: prop_types_1.default.shape({
        actions: Actions_1.default.propTypes.actions,
        dropdownProps: Actions_1.default.propTypes.dropdownProps,
        onSelect: Actions_1.default.propTypes.onSelect,
    }),
    expandAll: prop_types_1.default.oneOfType([
        prop_types_1.default.node,
        prop_types_1.default.shape({
            onClick: prop_types_1.default.func,
            isAllExpanded: prop_types_1.default.bool,
            isDisabled: prop_types_1.default.bool,
            buttonProps: prop_types_1.default.shape((_a = {},
                _a[prop_types_1.default.string] = prop_types_1.default.object,
                _a)),
        }),
    ]),
    useMobileLayout: prop_types_1.default.bool,
};
exports.default = PrimaryToolbar;
